feat(plugin): Title Cleaner OST

nrth3rnlb opened the pull request:
Because yesterday I foolishly deleted a branch, which led to the closure of [this pull request](https://github.com/metabrainz/picard-plugins/pull/399), I have revised the plugin somewhat and am hereby submitting a new pull request.

nrth3rnlb committed:
feat(plugin): Title Cleaner OST

Because yesterday I foolishly deleted a branch, which led to the closure of [this pull request](https://github.com/metabrainz/picard-plugins/pull/399), I have revised the plugin somewhat and am hereby submitting a new pull request.

Sophist-UK commented:
Not exactly a procedure, but it does help them know that you are finished tweaking and they can review it - otherwise that might not be obvious.

nrth3rnlb commented:
I assumed that the pull request itself was already a request for review.

nrth3rnlb committed:
fix: Corrected the names of two configuration parameters and replace string literals with constants

Fortunately, this would only have become problematic if an upstream action had required a fallback.

nrth3rnlb committed:
feat(ui): Update Title Cleaner OST configuration UI and add live update options

nrth3rnlb committed:
feat: Update Title Cleaner OST plugin to v1.3.0

- 1.3.0 (a58c7c4)
- fix: album process logic (16d3363)
- feat: enhance configuration options and UI (acdd0bd)

nrth3rnlb committed:
feat: Update Title Cleaner OST plugin to v1.3.1

- 1.3.1 (ef86848)
- fix: remove unused imports and simplify checkbox state handling (a624bac)

nrth3rnlb committed:
feat: Update Title Cleaner OST plugin to v1.3.2

- 1.3.2 (4524cff)
- fix(logic): Changes to the title will also be made if no checkbox is selected (4347de0)
- fix(logging): add debug log for title_cleaner_ost function (e422cd9)

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+Removes soundtrack-related information from album titles using regex.
+"""
+
+__version__ = "1.3.2"
```
Is there a specific point to separating out the version like this?

nrth3rnlb replied:
No, not really. The plugins are (still) in their own repositories. I probably went a bit overboard, which unnecessarily increased the complexity. I will reorganise that.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+from .ui_title_cleaner_ost_config import Ui_RemoveReleaseTitleOstIndicatorSettings
+
+# No longer used, but kept for config compatibility
+# ONLY_SOUNDTRACK = "title_cleaner_ost_only_soundtrack"
```
This is the first version of this plugin so there is no need for backwards compatibility.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+OST_REGEX = "title_cleaner_ost_regex"
+LIVE_UPDATES = "title_cleaner_ost_live_updates"
+APPLY_OPTIONS = "title_cleaner_ost_apply_options"
+SCHEMA_VERSION = "title_cleaner_ost_schema_version"
```
I do not understand the purpose of this. Aside from defining the setting, there is no use of this setting.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+
+def get_setting_with_default(key, default) -> Any:
+    """Helper to get a setting with a default fallback."""
+    return config.setting[key] if key in config.setting else default
```
Settings get a default as standard. This should be removed.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+    DEFAULT_REGEX = r'(\s*(?:(?::|：|∶|-|–|—|\(|\[)\s*)?(\b(?:Original|Album|Movie|Motion|Picture|Soundtrack|Score|OST|Music|Edition|Inspired|by|from|the|TV|Series|Video|Game|Film|Show)\b)+(?:\)|\])?\s*)+$'
+    DEFAULT_WHITELIST = ""
+
+    APPLY_OPTIONS_SCHEMA_VERSION = 1
```
I do not understand the purpose of this. Aside from defining the setting, there is no use of this setting.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+        TextOption("setting", OST_WHITELIST, DEFAULT_WHITELIST),
+        BoolOption("setting", LIVE_UPDATES, False),
+        ListOption("setting", APPLY_OPTIONS, DEFAULT_APPLY_OPTIONS),
+        IntOption("setting", SCHEMA_VERSION, APPLY_OPTIONS_SCHEMA_VERSION),
```
I do not understand the purpose of this. Aside from defining the setting, there is no use of this setting.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+
+    log.debug("%s: Processing album title with regex '%s'", PLUGIN_NAME, regex)
+
+    if "album" in metadata:
```
If you are changing `album`, then you should also change `albumsort`.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+        album_title = self.ui.test_input.text().strip()
+        whitelist = self.ui.whitelist_text.toPlainText()
+
+        # Normalise whitelist titles using Unicode NFC normalization
```
I wonder whether whitelisting titles by exact text match is a good idea. If the text changes on the server, then a pre-existing whitelist entry will no longer work.

Would it be better to whitelist Release_MBIDs and / or ReleaseGroup_MBIDs and/or instead allow users to define a regex for each line in the whitelist?

From a UI perspective, perhaps the user should be able to right-click on the album and add the album to the whitelist rather than having to go into options to add it. And if you are going to use MBIDs this would be a much easier way to whitelist them.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+
+        if should_process:
+            try:
+                compiled_regex = re.compile(regex, flags=re.IGNORECASE)
```
Why are you compiling this regex here - compiling this seems to be a UI action? And you definitely don't want to be compiling this once per album.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,246 @@
+        album_title = self.ui.test_input.text().strip()
+        whitelist = self.ui.whitelist_text.toPlainText()
+
+        # Normalise whitelist titles using Unicode NFC normalization
+        whitelist_titles = [
+            unicodedata.normalize('NFC', line.strip()).lower()
+            for line in whitelist.splitlines() if line.strip()
+        ]
+
+        # Normalise album title for comparison
+        normalized_title = unicodedata.normalize('NFC', album_title).lower()
```
I agree - you should provide a standardised function to unicode normalise strings in the exact same way.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+                log.debug(PLUGIN_NAME + ": Preview error: %s", e)
+                self.ui.test_output.setText(f"Regex error: {e}")
+        else:
+            self.ui.test_output.setText("Invalid regex pattern")
```
Does the user need reminding what pattern is invalid?

And does this need a log.debug or a log.warning?

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+                new_title = ' '.join(new_title.split()).strip()
+                self.ui.test_output.setText(new_title)
+            except Exception as e:
+                log.debug(PLUGIN_NAME + ": Preview error: %s", e)
```
log.debug or log.warning?

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+                self.ui.test_output.setText(new_title)
+            except Exception as e:
+                log.debug(PLUGIN_NAME + ": Preview error: %s", e)
+                self.ui.test_output.setText(f"Regex error: {e}")
```
From a UI perspective, error messages should be distinguishable from valid output by e.g. being prefixed with e.g. "ERROR: " or using e.g. coloured box border or coloured text etc.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+        """Resets the regex to the default pattern."""
+        self.ui.regex_pattern.setPlainText(self.DEFAULT_REGEX)
+
+    def validate_regex_pattern(self) -> bool:
```
It probably shouldn't be done here, but an invalid regex pattern should have a UI highlight (red text, red box border etc.) to indicate that it is invalid.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+
+    APPLY_OPTIONS_SCHEMA_VERSION = 1
+
+    DEFAULT_APPLY_OPTIONS: List[Dict[str, Any]] = [
```
Similarly, I do not understand the point of saving this as a config option when there is no UI for the user to be able to change it, and if the user is not going to change it then couldn't it just be a CLASS constant?

But if the User is going to be able to change it, then it will need a user friendly UI.

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+                    should_process = True
+                    continue
+    except AttributeError as e:
+        log.error("%s: Error processing apply options: %s", PLUGIN_NAME, e)
```
How will the user know what exactly is in error if you don't give them `option` in the error message?

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -0,0 +1,376 @@
+                new_title = compiled_regex.sub('', album_title)
+                # Normalise whitespace and strip leading/trailing whitespace
+                new_title = ' '.join(new_title.split()).strip()
+                metadata["album"] = new_title
```
Should there be a UI indicator for the album in the main Picard screen to show that it has been modified by this plugin?

P.S. I have no idea how you would show it (unless you add a unicode character to the end of the album title and remove it when saving).

Sophist-UK commented on plugins/title_cleaner_ost/ui_title_cleaner_ost_config.py:
```diff
@@ -0,0 +1,143 @@
+# run again.  Do not edit this file unless you know what you are doing.
+
+
+from PyQt5 import QtCore, QtGui, QtWidgets
```
Since this is an automatically generated .py file (from the .ui xml file), this copilot comment should IMO be ignored.

Sophist-UK reviewed (changes requested):
I am not someone who can approve this PR, but in the interests of helping you I am providing my own views in the hope that approvers would have the same comments and I am giving them to you earlier.

nrth3rnlb committed:
feat: Update Title Cleaner OST plugin to v1.3.3

- chore(init): bump PLUGIN_VERSION to 1.3.3 (4c38dd5)
- fix(init): remove __version__ and set PLUGIN_VERSION directly (905a4d5)
- 1.3.2 (4524cff)
- fix(logic): Changes to the title will also be made if no checkbox is selected (4347de0)
- fix(logging): add debug log for title_cleaner_ost function (e422cd9)

Sophist-UK commented on plugins/title_cleaner_ost/__init__.py:
```diff
@@ -5,8 +5,6 @@
 Removes soundtrack-related information from album titles using regex.
 """
 
-__version__ = "1.3.2"
```
verison 1.0.0 or v0.1.

nrth3rnlb committed:
Merge branch 'metabrainz:2.0' into title-cleaner-ost-plugin

nrth3rnlb committed:
Modifiziert von www.SourceFiles.app

nrth3rnlb committed:
chore: remove commented-out code

nrth3rnlb committed:
chore: fix version

nrth3rnlb committed:
chore: unnecessary constant removed

nrth3rnlb committed:
Merge branch 'title-cleaner-ost-plugin' of git@github.com:nrth3rnlb/picard-plugins.git

nrth3rnlb committed:
chore: remove obsolete APPLY_OPTIONS_SCHEMA_VERSION constant

nrth3rnlb committed:
Merge branch 'title-cleaner-ost-plugin' of git@github.com:nrth3rnlb/picard-plugins.git

nrth3rnlb committed:
refactor: unnecessarily complex access to material configuration removed

nrth3rnlb committed:
Merge branch 'title-cleaner-ost-plugin' of git@github.com:nrth3rnlb/picard-plugins.git

nrth3rnlb committed:
chore: remove unused APPLY_OPTIONS_SCHEMA_VERSION constant

Sophist-UK commented:
I am sorry that you are finding this so difficult - because it really isn't. 

The AI is giving you a code review - and this is really no different to a human code review from e.g. me or Zas or Philipp. All you need to do is to consider whether each of the suggestions is valid and reasonable or not and either follow the suggestion, or do something similar or add a comment stating why you are rejecting the review comment.

These review comments are teamwork contributions to help you to make a better plugin that has better code, is more robust and gives a better experience to users. They are not harsh or personal criticisms - so please take them in the helpful way that they are intended.

The tests are repo wide tests that run on every plugin to check the basics. The tests are self documenting and they are in  the repo, and so you can read them yourself to work out why your plugin(s) are not passing the tests. (And it may not be your plugin that is failing - it could be a different plugin.

As for processes which are not transparent, I am really not sure what you are referring to here. 

The process is a simple one:

1. You submit a PR - when you have finished development and have no more changes to make yourself, you mark it as final rather than draft and request reviews from the project's maintainers.

2. You mark the PR as draft again, respond to any review comments as described above, mark it as final again and request reviews again.

But as stated elsewhere you need to be patient because both Picard and Picard Plugins repos are managed by volunteers who have day jobs and lives and so cannot necessarily get to reviewing your submissions that quickly.  

P.S. When you lose patience and close the PR, you are trashing the time I have invested in trying to help you, and it feels to me like a slap in the face. Is this really the impression you want to give?
